sys: Modify ATOMIC_BITMAP_SIZE macro: num_bits = 0

Burt-Silverman opened the pull request:
The macro returned the wrong value for any case in which num_bits<=0, due to a signed/unsigned division. The num_bits=0 case was hit using static analysis in a Nordic Semiconductor SDK case.

Fixes #66356

andyross reviewed (commented):
`ROUND_UP(num_bits, ATOMIC_BITS)/ATOMIC_BITS` would be clearer IMHO

Burt-Silverman commented:
I like it. I will try to amend my commit and take it from there.

Burt-Silverman committed:
sys: Modify ATOMIC_BITMAP_SIZE macro: num_bits = 0

    The macro returned the wrong value for any case in which
    num_bits<=0, due to a signed/unsigned division. The num_bits=0 case was
    hit using static analysis in a Nordic Semiconductor SDK case.

Signed-off-by: Burt Silverman <burtms@gmail.com>

Burt-Silverman commented:
Unfortunately I had no clue how to modify the first commit. I was trying to do that back on my desktop and I couldn't push the change...Now after trying to do it directly on github it looks like I have 2 commits in this pull request and that can't be correct and I don't know what to do. Time to get some sleep. Any clues appreciated, or I'll try to read the github docs before giving up.

andyross commented:
Just make your changes, `git commit --amend atomic.h` (or recreate the commit from scratch, or `git rebase -i`, or any other technique you want to get the series of commits into exactly the form you want).  Then, when you push, use `git push -f` because you're not pushing a clean "fast forward" from the pre-existing ref.  Force pushing new versions of pull requests is the normal paradigm for Zephyr, and most other projects.

Github's native flow fits better with amending PR's during review, but that creates readability problems with understanding why changes were made (it also tends to break bisectability pretty badly).

Burt-Silverman commented:
It looks like I committed the ROUND_UP modification with the Github native flow prior to seeing your last comment, Andy. So, will that work for this one time?

andyross reviewed (approved):
This looks great, thanks.   You can always fix things up with a force push.  Though do be aware that if there's a lot of context to previous versions of a big PR, github will sometimes "lose" the commits if they aren't otherwise referenced (by another branch somewhere, etc..), which leads to some loss of informat if there is a lot of review chatter; you see the comments but not the code it was talking about.

Obviously not a problem here, but something to keep in mind.  I try to leave old PR source branches around in my repos for that reason.

Burt-Silverman commented:
I decided to go ahead and force push so there is only one commit. Sorry to be such a newbie to contributing; I contributed once to a project but that one (fdio) used gerrit.

Burt-Silverman commented:
I had forgotten to open a bug report previously. I have done that now: it is #66356.

Burt-Silverman commented:
Chris, how is this change?:

diff --git a/tests/kernel/common/src/atomic.c b/tests/kernel/common/src/atomic.c
index af5273ed25..a198dcc0bf 100644
--- a/tests/kernel/common/src/atomic.c
+++ b/tests/kernel/common/src/atomic.c
@@ -27,6 +27,25 @@ static struct k_thread thread[THREADS_NUM];
 
 atomic_t total_atomic;
 
+/*
+ * for integer n bits; n < -sizeof(atomic_t)*8, bytesize (in bits) = 8,
+ * wordsize (in bits) is 64 or 32, then:
+ * arraysize = (1L<<wordsize)/bytesize - n/wordsize*(wordsize/bytesize)
+ *           = (1L<<(wordsize-3)) - n/wordsize*(wordsize/bytesize)
+ * So arraysize is huge in those cases; that was not my intention.
+ * However, n_bits should be >0; possibly 0 for clang-tidy static analysis.
+ */
+atomic_t array0[ATOMIC_BITMAP_SIZE(-401)];
+atomic_t array1[ATOMIC_BITMAP_SIZE(0)];
+atomic_t array2[ATOMIC_BITMAP_SIZE(1)];
+atomic_t array3[ATOMIC_BITMAP_SIZE(66)];
+
+BUILD_ASSERT(sizeof array0 == (1L<<ATOMIC_WORD(61,29))
+                               - 401/ATOMIC_WORD(64,32)*ATOMIC_WORD(8,4));
+BUILD_ASSERT(sizeof array1 == 0);
+BUILD_ASSERT(sizeof array2 / sizeof(atomic_t) == 1);
+BUILD_ASSERT(sizeof array3 / sizeof(atomic_t) == ATOMIC_WORD(2, 3));
+
 /**
  * @addtogroup kernel_common_tests
  * @{

Burt-Silverman commented:
But changing array0 through array4 to static, e.g.
static atomic_t array0[ATOMIC_BITMAP_SIZE(-401)];
